'wire_core' error after April Update

I am getting an error after the latest update in April
```
[ERROR] [1621244522.215041341]: While parsing '/home/aito-dm/catkin_ws/src/wire/wire_core/models/world_object_models.xml': 

Could not find attribute 'value' of element 'prior_new'
Could not find attribute 'value' of element 'prior_existing'
Could not find attribute 'value' of element 'prior_clutter'
In class description for 'object': Could not find attribute 'dimensions' of element 'pnew'

In class description for 'object': Could not find attribute 'domain_size' of element 'pnew'

In class description for 'object': Could not find attribute 'domain_size' of element 'pnew'

In class description for 'object': Could not find attribute 'domain_size' of element 'pnew'


[wire_server-2] process has died [pid 14235, exit code -11, cmd /home/aito-dm/catkin_ws/devel/lib/wire_core/wire_server __name:=wire_server __log:=/root/.ros/log/21772a22-b6f4-11eb-a603-6c6a77dd951e/wire_server-2.log].
```

--- a/wire_core/CMakeLists.txt
+++ b/wire_core/CMakeLists.txt
@@ -35,8 +35,8 @@ catkin_package(
 include_directories(include ${catkin_INCLUDE_DIRS} ${TinyXML2_INCLUDE_DIRS})
 
 ## Build
-add_compile_options(-std=c++0x) # C++ 11
 add_compile_options(-Wreturn-type) # Return type checking
+add_compile_options(-Wreorder)
 
 
 add_library(wire
@@ -87,6 +87,10 @@ add_executable(dummy_publisher src/dummy_evidence_publisher.cpp)
 target_link_libraries(dummy_publisher wire)
 
 if(CATKIN_ENABLE_TESTING)
+  add_library(test_plugin EXCLUDE_FROM_ALL SHARED test/test_plugin.cpp)
+  target_link_libraries(test_plugin wire)
+
   catkin_add_gtest(test_parser test/test_parser.cpp)
   target_link_libraries(test_parser wire)
+  add_dependencies(test_parser test_plugin)
 endif()

--- a/wire_core/CMakeLists.txt
+++ b/wire_core/CMakeLists.txt
@@ -85,3 +85,8 @@ target_link_libraries(wire_server wire)
 
 add_executable(dummy_publisher src/dummy_evidence_publisher.cpp) 
 target_link_libraries(dummy_publisher wire)
+
+if(CATKIN_ENABLE_TESTING)
+  catkin_add_gtest(test_parser test/test_parser.cpp)
+  target_link_libraries(test_parser wire)
+endif()

--- a/wire_core/include/wire/core/IStateEstimator.h
+++ b/wire_core/include/wire/core/IStateEstimator.h
@@ -103,7 +103,7 @@ public:
      * @param b The boolean value
      * @return Returns true if the parameter was known to the estimator; false otherwise
      */
-    virtual bool setParameter(const std::string& param, bool b) {
+    virtual bool setParameter(const std::string& /*param*/, bool /*b*/) {
         return false;
     }
 
@@ -113,7 +113,7 @@ public:
      * @param v The float value
      * @return Returns true if the parameter was known to the estimator; false otherwise
      */
-    virtual bool setParameter(const std::string& param, double v) {
+    virtual bool setParameter(const std::string& /*param*/, double /*v*/) {
         return false;
     }
 
@@ -123,7 +123,7 @@ public:
      * @param s The string value
      * @return Returns true if the parameter was known to the estimator; false otherwise
      */
-    virtual bool setParameter(const std::string& param, const std::string& s) {
+    virtual bool setParameter(const std::string& /*param*/, const std::string& /*s*/) {
         return false;
     }
 };

--- a/wire_core/include/wire/util/ObjectModelParser.h
+++ b/wire_core/include/wire/util/ObjectModelParser.h
@@ -26,7 +26,7 @@ class ObjectModelParser {
 
 public:
 
-    ObjectModelParser(const std::string& filename);
+    ObjectModelParser(const std::string& filename, const std::string& plugin_attr="plugin");
 
     virtual ~ObjectModelParser();
 

--- a/wire_core/models/world_object_models.xml
+++ b/wire_core/models/world_object_models.xml
@@ -1,35 +1,36 @@
+<?xml version="1.0"?>
 <knowledge>
 
     <prior_new value="0.14" />
     <prior_existing value="0.14" />
     <prior_clutter value="0.72" />
 
-    <object_class name="object"> 
-    
+    <object_class name="object">
+
         <behavior_model attribute="position" model="wire_state_estimators/PositionEstimator">
-            <pnew type="uniform" dimensions="3" density="0.0001" />            
+            <pnew type="uniform" dimensions="3" density="0.0001" />
             <pclutter type="uniform" dimensions="3" density="0.0001" />
 
-            <param name="max_acceleration" value="8" />                
-            <param name="kalman_timeout" value="1" />                
-            <param name="fixed_pdf_cov" value="0.008" />                
+            <param name="max_acceleration" value="8" />
+            <param name="kalman_timeout" value="1" />
+            <param name="fixed_pdf_cov" value="0.008" />
         </behavior_model>
 
         <behavior_model attribute="color" model="wire_state_estimators/DiscreteEstimator">
-            <pnew type="discrete" domain_size="100" />            
+            <pnew type="discrete" domain_size="100" />
             <pclutter type="discrete" domain_size="100" />
-        </behavior_model> 
+        </behavior_model>
 
         <behavior_model attribute="class_label" model="wire_state_estimators/DiscreteEstimator">
-            <pnew type="discrete" domain_size="100" />            
+            <pnew type="discrete" domain_size="100" />
             <pclutter type="discrete" domain_size="100" />
         </behavior_model>
 
         <behavior_model attribute="shape" model="wire_state_estimators/DiscreteEstimator">
-            <pnew type="discrete" domain_size="10" />            
+            <pnew type="discrete" domain_size="10" />
             <pclutter type="discrete" domain_size="10" />
         </behavior_model>
 
     </object_class>
-    
+
 </knowledge>

--- a/wire_core/package.xml
+++ b/wire_core/package.xml
@@ -23,5 +23,9 @@
   <depend>tinyxml2</depend>
   <depend>wire_msgs</depend>
 
+<export>
+    <wire_core plugin_test="${prefix}/test/test_plugins.xml" />
+  </export>
+
 </package>
 

--- a/wire_core/src/core/Property.cpp
+++ b/wire_core/src/core/Property.cpp
@@ -10,7 +10,7 @@
 
 namespace mhf {
 
-Property::Property(const Attribute& attribute, const IStateEstimator& bm, const ObjectID& object_id)
+Property::Property(const Attribute& attribute, const IStateEstimator& bm, const ObjectID& /*object_id*/)
     : time_(0), attribute_(attribute), estimator_(bm.clone()) {
 }
 
@@ -69,7 +69,7 @@ double Property::getLikelihood(const pbl::PDF& pdf) const {
     return estimator_->getValue().getLikelihood(pdf);
 }
 
-std::string Property::toString(const std::string& prefix) const {
+std::string Property::toString(const std::string& /*prefix*/) const {
     return estimator_->getValue().toString();
 }
 

--- a/wire_core/src/logic/HypothesesTree.cpp
+++ b/wire_core/src/logic/HypothesesTree.cpp
@@ -281,7 +281,7 @@ void HypothesisTree::normalizeProbabilities() {
     root_->calculateBranchProbabilities();
 }
 
-void HypothesisTree::pruneTree(const Time& timestamp) {
+void HypothesisTree::pruneTree(const Time& /*timestamp*/) {
     //return;
 
 

--- a/wire_core/src/models/FixedState.cpp
+++ b/wire_core/src/models/FixedState.cpp
@@ -27,10 +27,10 @@ FixedState* FixedState::clone() const {
     return new FixedState(*this);
 }
 
-void FixedState::update(const pbl::PDF& z, const mhf::Time& time) {
+void FixedState::update(const pbl::PDF& z, const mhf::Time& /*time*/) {
 }
 
-void FixedState::propagate(const mhf::Time& time) {
+void FixedState::propagate(const mhf::Time& /*time*/) {
 }
 
 void FixedState::reset() {

--- a/wire_core/src/util/ObjectModelParser.cpp
+++ b/wire_core/src/util/ObjectModelParser.cpp
@@ -57,13 +57,13 @@ bool ObjectModelParser::getAttributeValue(const tinyxml2::XMLElement* elem, stri
 bool ObjectModelParser::getAttributeValue(const tinyxml2::XMLElement* elem, string att_name, double& att_value, std::stringstream& error) {
     if (!elem) return false;
 
-    const char* value = elem->Attribute(att_name.c_str(), std::to_string(att_value).c_str());
-
+    const char* value = elem->Attribute(att_name.c_str());
     if (!value) {
         error << "Could not find attribute '" << att_name << "' of element '" << elem->Value() << "'" << endl;
         return false;
     }
 
+    att_value = std::atof(value);
     return true;
 }
 

--- a/wire_core/src/util/ObjectModelParser.cpp
+++ b/wire_core/src/util/ObjectModelParser.cpp
@@ -12,12 +12,10 @@
 #include <problib/pdfs/Uniform.h>
 #include <problib/pdfs/PMF.h>
 
-using namespace std;
-
 namespace mhf {
 
-ObjectModelParser::ObjectModelParser(const std::string& filename) : filename_(filename),
-    object_model_loader_(new pluginlib::ClassLoader<IStateEstimator>("wire_core", "IStateEstimator")) {
+ObjectModelParser::ObjectModelParser(const std::string& filename, const std::string& plugin_attr) : filename_(filename),
+    object_model_loader_(new pluginlib::ClassLoader<IStateEstimator>("wire_core", "IStateEstimator", plugin_attr)) {
 
 }
 
@@ -30,23 +28,23 @@ std::string ObjectModelParser::getErrorMessage() const {
     return parse_errors_.str();
 }
 
-string ObjectModelParser::getPropertyValue(const tinyxml2::XMLElement* elem, string prop_name, double& value, std::stringstream& error, bool optional) {
+std::string ObjectModelParser::getPropertyValue(const tinyxml2::XMLElement* elem, std::string prop_name, double& value, std::stringstream& error, bool optional) {
     const tinyxml2::XMLElement* p = elem->FirstChildElement(prop_name.c_str());
     if (p) {
         return p->Attribute("value", std::to_string(value).c_str());
     }
     if (!optional) {
-        error << "Could not find property '" << prop_name << "' of element '" << elem->Value() << "'" << endl;
+        error << "Could not find property '" << prop_name << "' of element '" << elem->Value() << "'" << std::endl;
     }
     return "";
 }
 
-bool ObjectModelParser::getAttributeValue(const tinyxml2::XMLElement* elem, string att_name, string& att_value, std::stringstream& error) {
+bool ObjectModelParser::getAttributeValue(const tinyxml2::XMLElement* elem, std::string att_name, std::string& att_value, std::stringstream& error) {
     if (!elem) return false;
 
     const char* value = elem->Attribute(att_name.c_str());
     if (!value) {
-        error << "Could not find attribute '" << att_name << "' of element '" << elem->Value() << "'" << endl;
+        error << "Could not find attribute '" << att_name << "' of element '" << elem->Value() << "'" << std::endl;
         return false;
     }
 
@@ -54,12 +52,12 @@ bool ObjectModelParser::getAttributeValue(const tinyxml2::XMLElement* elem, stri
     return true;
 }
 
-bool ObjectModelParser::getAttributeValue(const tinyxml2::XMLElement* elem, string att_name, double& att_value, std::stringstream& error) {
+bool ObjectModelParser::getAttributeValue(const tinyxml2::XMLElement* elem, std::string att_name, double& att_value, std::stringstream& error) {
     if (!elem) return false;
 
     const char* value = elem->Attribute(att_name.c_str());
     if (!value) {
-        error << "Could not find attribute '" << att_name << "' of element '" << elem->Value() << "'" << endl;
+        error << "Could not find attribute '" << att_name << "' of element '" << elem->Value() << "'" << std::endl;
         return false;
     }
 
@@ -67,36 +65,36 @@ bool ObjectModelParser::getAttributeValue(const tinyxml2::XMLElement* elem, stri
     return true;
 }
 
-bool ObjectModelParser::hasAttributeValue(const tinyxml2::XMLElement* elem, string att_name, string att_value) {
+bool ObjectModelParser::hasAttributeValue(const tinyxml2::XMLElement* elem, std::string att_name, std::string att_value) {
     if (!elem) return false;
 
     const char* value = elem->Attribute(att_name.c_str());
     if (!value) return false;
 
-    string valueStr = value;
+    std::string valueStr = value;
     return (att_value == valueStr);
 }
 
 pbl::PDF* ObjectModelParser::parsePDF(const tinyxml2::XMLElement* pdf_elem, std::stringstream& error) {
     const char* pdf_type = pdf_elem->Attribute("type");
     if (pdf_type) {
-        if (string(pdf_type) == "uniform") {
+        if (std::string(pdf_type) == "uniform") {
             double dim = 0;
             double density = 0;
             if (getAttributeValue(pdf_elem, "dimensions", dim, error)
                     && getAttributeValue(pdf_elem, "density", density, error)) {
                 return new pbl::Uniform((int)dim, density);
             }
-        } else if (string(pdf_type) == "discrete") {
+        } else if (std::string(pdf_type) == "discrete") {
             double domain_size;
             if (getAttributeValue(pdf_elem, "domain_size", domain_size, error)) {
                 return new pbl::PMF((int)domain_size);
             }
         } else {
-            error << "Unknown pdf type: " << pdf_type << endl;
+            error << "Unknown pdf type: " << pdf_type << std::endl;
         }
     } else {
-        error << "PDF specification should contain 'type' attribute" << endl;
+        error << "PDF specification should contain 'type' attribute" << std::endl;
     }
     return 0;
 }
@@ -104,7 +102,7 @@ pbl::PDF* ObjectModelParser::parsePDF(const tinyxml2::XMLElement* pdf_elem, std:
 bool ObjectModelParser::parseStateEstimator(ClassModel* obj_model, const tinyxml2::XMLElement* elem, std::stringstream& error) {
 
     // check behavior model's attribute and model type
-    string attribute_name, model_type;
+    std::string attribute_name, model_type;
     if (!getAttributeValue(elem, "attribute", attribute_name, error)
             | !getAttributeValue(elem, "model", model_type, error)) {
         return false;
@@ -128,9 +126,9 @@ bool ObjectModelParser::parseStateEstimator(ClassModel* obj_model, const tinyxml
     IStateEstimator* estimator;
 
     if (object_model_loader_->isClassAvailable(model_type)) {
-        estimator = object_model_loader_->createClassInstance(model_type)->clone();
+        estimator = object_model_loader_->createUnmanagedInstance(model_type)->clone();
     } else {
-        error << "Unknown model: " << model_type << endl;
+        error << "Unknown model: " << model_type << std::endl;
         return false;
     }
 
@@ -146,26 +144,26 @@ bool ObjectModelParser::parseStateEstimator(ClassModel* obj_model, const tinyxml
 
             bool set_param_ok = true;
             if (param->QueryDoubleAttribute("value", &v_double) == tinyxml2::XML_SUCCESS) {
-                set_param_ok = estimator->setParameter(string(param_name), v_double);
+                set_param_ok = estimator->setParameter(std::string(param_name), v_double);
             } else if (param->QueryIntAttribute("value", &v_int) == tinyxml2::XML_SUCCESS) {
-                set_param_ok = estimator->setParameter(string(param_name), (double)v_int);
+                set_param_ok = estimator->setParameter(std::string(param_name), (double)v_int);
             } else if (param->QueryBoolAttribute("value", &v_bool) == tinyxml2::XML_SUCCESS) {
-                set_param_ok = estimator->setParameter(string(param_name), v_bool);
+                set_param_ok = estimator->setParameter(std::string(param_name), v_bool);
             } else {
                 const char* v_str = param->Attribute("value");
                 if (v_str) {
-                    set_param_ok = estimator->setParameter(string(param_name), string(v_str));
+                    set_param_ok = estimator->setParameter(std::string(param_name), std::string(v_str));
                 } else {
-                    error << "State estimator parameters should always have a 'name' and 'value' attribute." << endl;
+                    error << "State estimator parameters should always have a 'name' and 'value' attribute." << std::endl;
                 }
             }
 
             if (!set_param_ok) {
-                error << "Unknown parameter for estimator '" << model_type << "': " << param_name << endl;
+                error << "Unknown parameter for estimator '" << model_type << "': " << param_name << std::endl;
             }
 
         } else {
-            error << "State estimator parameters should always have a 'name' and 'value' attribute." << endl;
+            error << "State estimator parameters should always have a 'name' and 'value' attribute." << std::endl;
         }
 
         param = param->NextSiblingElement("param");
@@ -184,7 +182,7 @@ bool ObjectModelParser::parseStateEstimator(ClassModel* obj_model, const tinyxml
             return false;
         }
     } else {
-        error << "Estimator specification does not contain 'pnew'." << endl;
+        error << "Estimator specification does not contain 'pnew'." << std::endl;
         return false;
     }
 
@@ -199,7 +197,7 @@ bool ObjectModelParser::parseStateEstimator(ClassModel* obj_model, const tinyxml
             return false;
         }
     } else {
-        error << "Estimator specification does not contain 'pclutter'." << endl;
+        error << "Estimator specification does not contain 'pclutter'." << std::endl;
         return false;
     }
 
@@ -208,12 +206,12 @@ bool ObjectModelParser::parseStateEstimator(ClassModel* obj_model, const tinyxml
     return true;
 }
 
-bool ObjectModelParser::getStateEstimatorParameter(const tinyxml2::XMLElement* elem, const string& param_name, double& value) {
+bool ObjectModelParser::getStateEstimatorParameter(const tinyxml2::XMLElement* elem, const std::string& param_name, double& value) {
 
     const tinyxml2::XMLElement* param = elem->FirstChildElement("param");
     while (param) {
         const char* v = param->Attribute("name");
-        if (v && (string)v == param_name) {
+        if (v && (std::string)v == param_name) {
             param->Attribute("value", std::to_string(value).c_str());
             return true;
         }
@@ -229,7 +227,7 @@ bool ObjectModelParser::parse(KnowledgeDatabase& knowledge_db) {
     doc.LoadFile(filename_.c_str());
 
     if (doc.Error()) {
-        ROS_ERROR_STREAM("While parsing '" << filename_ << "': " << endl << endl << doc.ErrorStr() << " at line " << doc.ErrorLineNum());
+        ROS_ERROR_STREAM("While parsing '" << filename_ << "': " << std::endl << std::endl << doc.ErrorStr() << " at line " << doc.ErrorLineNum());
         return false;
     }
 
@@ -242,7 +240,7 @@ bool ObjectModelParser::parse(KnowledgeDatabase& knowledge_db) {
             knowledge_db.setPriorNew(prior_new);
         }
     } else {
-        parse_errors_ << "Knowledge file does not contain 'prior_new'" << endl;
+        parse_errors_ << "Knowledge file does not contain 'prior_new'" << std::endl;
     }
 
     double prior_existing;
@@ -252,7 +250,7 @@ bool ObjectModelParser::parse(KnowledgeDatabase& knowledge_db) {
             knowledge_db.setPriorExisting(prior_existing);
         }
     } else {
-        parse_errors_ << "Knowledge file does not contain 'prior_existing'" << endl;
+        parse_errors_ << "Knowledge file does not contain 'prior_existing'" << std::endl;
     }
 
     double prior_clutter;
@@ -262,7 +260,7 @@ bool ObjectModelParser::parse(KnowledgeDatabase& knowledge_db) {
             knowledge_db.setPriorClutter(prior_clutter);
         }
     } else {
-        parse_errors_ << "Knowledge file does not contain 'prior_clutter'" << endl;
+        parse_errors_ << "Knowledge file does not contain 'prior_clutter'" << std::endl;
     }
 
 
@@ -273,12 +271,12 @@ bool ObjectModelParser::parse(KnowledgeDatabase& knowledge_db) {
     while(class_element) {
         ClassModel* class_model = 0;
 
-        string model_name;
+        std::string model_name;
         getAttributeValue(class_element, "name", model_name, parse_errors_);
 
-        cout << "Parsing model for class " << model_name << endl;
+        std::cout << "Parsing model for class " << model_name << std::endl;
 
-        string base_class = "";
+        std::string base_class = "";
         const char* value = class_element->Attribute("base");
 
         if (value) {
@@ -290,7 +288,7 @@ bool ObjectModelParser::parse(KnowledgeDatabase& knowledge_db) {
                 class_model = new ClassModel(*base_model);
                 class_model->setModelName(model_name);
             } else {
-                parse_errors_ << "Error in class definition of '" << model_name << "': unknown base class '" << base_class << "'." << endl;
+                parse_errors_ << "Error in class definition of '" << model_name << "': unknown base class '" << base_class << "'." << std::endl;
                 class_model = new ClassModel(model_name);
             }
         } else {
@@ -300,15 +298,15 @@ bool ObjectModelParser::parse(KnowledgeDatabase& knowledge_db) {
         // parse properties
         const tinyxml2::XMLElement* prop = class_element->FirstChildElement();
         while(prop) {
-            string prop_name = prop->Value();
+            std::string prop_name = prop->Value();
             if (prop_name == "behavior_model") {
-                stringstream bh_errors;
+                std::stringstream bh_errors;
                 parseStateEstimator(class_model, prop, bh_errors);
                 if (bh_errors.str() != "") {
-                    parse_errors_ << "In class description for '" << class_model->getModelName() << "': " << bh_errors.str() << endl;
+                    parse_errors_ << "In class description for '" << class_model->getModelName() << "': " << bh_errors.str() << std::endl;
                 }
             } else {
-                parse_errors_ << "In class description for '" << class_model->getModelName() << "': Unknown class property: '" << prop_name << "'" << endl;
+                parse_errors_ << "In class description for '" << class_model->getModelName() << "': Unknown class property: '" << prop_name << "'" << std::endl;
             }
             prop = prop->NextSiblingElement();
         }

--- a/wire_core/test/test_parser.cpp
+++ b/wire_core/test/test_parser.cpp
@@ -1,0 +1,20 @@
+#include <gtest/gtest.h>
+
+#include <ros/package.h>
+
+#include <wire/util/ObjectModelParser.h>
+#include <wire/storage/KnowledgeDatabase.h>
+
+
+TEST(ObjectModelParser, parse){
+    std::string path = ros::package::getPath("wire_core");
+    path += "/models/world_object_models.xml";
+    mhf::ObjectModelParser parser(path);
+
+    ASSERT_TRUE(parser.parse(mhf::KnowledgeDatabase::getInstance()));
+}
+
+int main(int argc, char **argv){
+   testing::InitGoogleTest(&argc, argv);
+   return RUN_ALL_TESTS();
+}

--- a/wire_core/test/test_parser.cpp
+++ b/wire_core/test/test_parser.cpp
@@ -9,9 +9,10 @@
 TEST(ObjectModelParser, parse){
     std::string path = ros::package::getPath("wire_core");
     path += "/models/world_object_models.xml";
+    std::cout << "Parsing: " << path << std::endl;
     mhf::ObjectModelParser parser(path);
 
-    ASSERT_TRUE(parser.parse(mhf::KnowledgeDatabase::getInstance()));
+    ASSERT_TRUE(parser.parse(mhf::KnowledgeDatabase::getInstance())) << parser.getErrorMessage();
 }
 
 int main(int argc, char **argv){

--- a/wire_core/test/test_parser.cpp
+++ b/wire_core/test/test_parser.cpp
@@ -8,9 +8,9 @@
 
 TEST(ObjectModelParser, parse){
     std::string path = ros::package::getPath("wire_core");
-    path += "/models/world_object_models.xml";
+    path += "/test/test_world_object_models.xml";
     std::cout << "Parsing: " << path << std::endl;
-    mhf::ObjectModelParser parser(path);
+    mhf::ObjectModelParser parser(path, "plugin_test");
 
     ASSERT_TRUE(parser.parse(mhf::KnowledgeDatabase::getInstance())) << parser.getErrorMessage();
 }

--- a/wire_core/test/test_plugin.cpp
+++ b/wire_core/test/test_plugin.cpp
@@ -1,0 +1,103 @@
+/************************************************************************
+ *  Copyright (C) 2021 Eindhoven University of Technology (TU/e).       *
+ *  All rights reserved.                                                *
+ ************************************************************************
+ *  Redistribution and use in source and binary forms, with or without  *
+ *  modification, are permitted provided that the following conditions  *
+ *  are met:                                                            *
+ *                                                                      *
+ *      1.  Redistributions of source code must retain the above        *
+ *          copyright notice, this list of conditions and the following *
+ *          disclaimer.                                                 *
+ *                                                                      *
+ *      2.  Redistributions in binary form must reproduce the above     *
+ *          copyright notice, this list of conditions and the following *
+ *          disclaimer in the documentation and/or other materials      *
+ *          provided with the distribution.                             *
+ *                                                                      *
+ *  THIS SOFTWARE IS PROVIDED BY TU/e "AS IS" AND ANY EXPRESS OR        *
+ *  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED      *
+ *  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE  *
+ *  ARE DISCLAIMED. IN NO EVENT SHALL TU/e OR CONTRIBUTORS BE LIABLE    *
+ *  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR        *
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT   *
+ *  OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;     *
+ *  OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF       *
+ *  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT           *
+ *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE   *
+ *  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH    *
+ *  DAMAGE.                                                             *
+ *                                                                      *
+ *  The views and conclusions contained in the software and             *
+ *  documentation are those of the authors and should not be            *
+ *  interpreted as representing official policies, either expressed or  *
+ *  implied, of TU/e.                                                   *
+ ************************************************************************/
+
+#include "test_plugin.h"
+
+TestPlugin::TestPlugin() : pdf_(nullptr), bool_param_(false), double_param_(0), string_param_() {
+}
+
+TestPlugin::TestPlugin(const pbl::PDF& pdf) : pdf_(pdf.clone()) {
+}
+
+TestPlugin::TestPlugin(const TestPlugin& orig) : mhf::IStateEstimator(orig), pdf_(nullptr),
+    bool_param_(orig.bool_param_), double_param_(orig.double_param_), string_param_(orig.string_param_) {
+
+    if (orig.pdf_) {
+        pdf_ = orig.pdf_->clone();
+    }
+}
+
+TestPlugin::~TestPlugin() {
+    if (pdf_)
+        delete pdf_;
+}
+
+TestPlugin* TestPlugin::clone() const {
+    return new TestPlugin(*this);
+}
+
+void TestPlugin::update(const pbl::PDF& /*z*/, const mhf::Time& /*time*/) {
+}
+
+void TestPlugin::propagate(const mhf::Time& /*time*/) {
+}
+
+void TestPlugin::reset() {
+}
+
+const pbl::PDF& TestPlugin::getValue() const {
+    return *pdf_;
+}
+
+bool TestPlugin::setParameter(const std::string& param, bool b) {
+    if (param == "bool_param") {
+        bool_param_ = b;
+    } else {
+        return false;
+    }
+    return true;
+}
+
+bool TestPlugin::setParameter(const std::string &param, double v) {
+    if (param == "double_param") {
+        double_param_ = v;
+    } else {
+        return false;
+    }
+    return true;
+}
+
+bool TestPlugin::setParameter(const std::string& param, const std::string& s) {
+    if (param == "string_param") {
+        string_param_ = s;
+    } else {
+        return false;
+    }
+    return true;
+}
+
+#include <pluginlib/class_list_macros.h>
+PLUGINLIB_EXPORT_CLASS( TestPlugin, mhf::IStateEstimator )

--- a/wire_core/test/test_plugin.h
+++ b/wire_core/test/test_plugin.h
@@ -1,0 +1,113 @@
+/************************************************************************
+ *  Copyright (C) 2021 Eindhoven University of Technology (TU/e).       *
+ *  All rights reserved.                                                *
+ ************************************************************************
+ *  Redistribution and use in source and binary forms, with or without  *
+ *  modification, are permitted provided that the following conditions  *
+ *  are met:                                                            *
+ *                                                                      *
+ *      1.  Redistributions of source code must retain the above        *
+ *          copyright notice, this list of conditions and the following *
+ *          disclaimer.                                                 *
+ *                                                                      *
+ *      2.  Redistributions in binary form must reproduce the above     *
+ *          copyright notice, this list of conditions and the following *
+ *          disclaimer in the documentation and/or other materials      *
+ *          provided with the distribution.                             *
+ *                                                                      *
+ *  THIS SOFTWARE IS PROVIDED BY TU/e "AS IS" AND ANY EXPRESS OR        *
+ *  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED      *
+ *  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE  *
+ *  ARE DISCLAIMED. IN NO EVENT SHALL TU/e OR CONTRIBUTORS BE LIABLE    *
+ *  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR        *
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT   *
+ *  OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;     *
+ *  OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF       *
+ *  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT           *
+ *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE   *
+ *  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH    *
+ *  DAMAGE.                                                             *
+ *                                                                      *
+ *  The views and conclusions contained in the software and             *
+ *  documentation are those of the authors and should not be            *
+ *  interpreted as representing official policies, either expressed or  *
+ *  implied, of TU/e.                                                   *
+ ************************************************************************/
+
+#ifndef TEST_TEST_PLUGIN_H_
+#define TEST_TEST_PLUGIN_H_
+
+#include "wire/core/IStateEstimator.h"
+
+/**
+ * @brief TestPlugin is a simple wrapper such that probability density function can
+ * be threaded as state estimator, i.e., TestPlugin reflects a state but update,
+ * propagation and reset do not influence the state.
+ */
+class TestPlugin : public mhf::IStateEstimator {
+
+public:
+
+    TestPlugin();
+
+    TestPlugin(const pbl::PDF& pdf);
+
+    TestPlugin(const TestPlugin& orig);
+
+    virtual ~TestPlugin();
+
+    virtual TestPlugin* clone() const;
+
+    /**
+     * @brief Performs an update, but since the state is fixed, update will do nothing.
+     */
+    virtual void update(const pbl::PDF& z, const mhf::Time& time);
+
+    /**
+     * @brief Propagates the state, but since the state is fixed, propagate will do nothing.
+     */
+
+    virtual void propagate(const mhf::Time& time);
+
+    /**
+     * @brief Resets the state, but since the state is fixed, reset will do nothing.
+     */
+    virtual void reset();
+
+    const pbl::PDF& getValue() const;
+
+    /**
+     * @brief Set a boolean parameter of this state estimator
+     * @param param The parameter name
+     * @param b The boolean value
+     * @return Returns true if the parameter was known to the estimator; false otherwise
+     */
+    bool setParameter(const std::string& param, bool b);
+
+    /**
+     * @brief Set a real-valued parameter of this state estimator
+     * @param param The parameter name
+     * @param v The float value
+     * @return Returns true if the parameter was known to the estimator; false otherwise
+     */
+    bool setParameter(const std::string& param, double v);
+
+    /**
+     * @brief Set a string parameter of this state estimator
+     * @param param The parameter name
+     * @param s The string value
+     * @return Returns true if the parameter was known to the estimator; false otherwise
+     */
+    bool setParameter(const std::string& param, const std::string& s);
+
+protected:
+
+    pbl::PDF* pdf_;
+
+    bool bool_param_;
+    double double_param_;
+    std::string string_param_;
+
+};
+
+#endif /* TEST_TEST_PLUGIN_H_ */

--- a/wire_core/test/test_plugins.xml
+++ b/wire_core/test/test_plugins.xml
@@ -1,0 +1,6 @@
+<?xml version="1.0"?>
+<library path="lib/libtest_plugin">
+  <class name="wire_core/Test" type="TestPlugin" base_class_type="IStateEstimator">
+    <description>This is a test plugin.</description>
+  </class>
+</library>

--- a/wire_core/test/test_world_object_models.xml
+++ b/wire_core/test/test_world_object_models.xml
@@ -1,0 +1,21 @@
+<?xml version="1.0"?>
+<knowledge>
+
+    <prior_new value="0.14" />
+    <prior_existing value="0.14" />
+    <prior_clutter value="0.72" />
+
+    <object_class name="object">
+
+        <behavior_model attribute="position" model="wire_core/Test">
+            <pnew type="uniform" dimensions="3" density="0.0001" />
+            <pclutter type="uniform" dimensions="3" density="0.0001" />
+
+            <param name="bool_param" value="true" />
+            <param name="double_param" value="42.42" />
+            <param name="string_param" value="foo" />
+        </behavior_model>
+
+    </object_class>
+
+</knowledge>

--- a/wire_state_estimators/src/DiscreteFilter.cpp
+++ b/wire_state_estimators/src/DiscreteFilter.cpp
@@ -54,10 +54,10 @@ DiscreteFilter* DiscreteFilter::clone() const {
 	return new DiscreteFilter(*this);
 }
 
-void DiscreteFilter::propagate(const mhf::Time& time) {
+void DiscreteFilter::propagate(const mhf::Time& /*time*/) {
 }
 
-void DiscreteFilter::update(const pbl::PDF& z, const mhf::Time& time) {
+void DiscreteFilter::update(const pbl::PDF& z, const mhf::Time& /*time*/) {
 	assert(z.type() == pbl::PDF::DISCRETE);
 	const pbl::PMF* pmf = pbl::PDFtoPMF(z);
 	pmf_.update(*pmf);

--- a/wire_state_estimators/src/FixedState.cpp
+++ b/wire_state_estimators/src/FixedState.cpp
@@ -51,13 +51,13 @@ FixedState::~FixedState() {
 }
 
 FixedState* FixedState::clone() const {
-	return new FixedState(*this);
+    return new FixedState(*this);
 }
 
-void FixedState::update(const pbl::PDF& z, const mhf::Time& time) {
+void FixedState::update(const pbl::PDF& /*z*/, const mhf::Time& /*time*/) {
 }
 
-void FixedState::propagate(const mhf::Time& time) {
+void FixedState::propagate(const mhf::Time& /*time*/) {
 }
 
 void FixedState::reset() {

--- a/wire_state_estimators/src/FixedState.cpp
+++ b/wire_state_estimators/src/FixedState.cpp
@@ -43,11 +43,14 @@ FixedState::FixedState() {
 FixedState::FixedState(const pbl::PDF& pdf) : pdf_(pdf.clone()) {
 }
 
-FixedState::FixedState(const FixedState& orig) : mhf::IStateEstimator(orig), pdf_(orig.pdf_->clone()) {
+FixedState::FixedState(const FixedState& orig) : mhf::IStateEstimator(orig), pdf_(nullptr) {
+    if (orig.pdf_)
+        pdf_ = orig.pdf_->clone();
 }
 
 FixedState::~FixedState() {
-	delete pdf_;
+    if (pdf_)
+        delete pdf_;
 }
 
 FixedState* FixedState::clone() const {

--- a/wire_state_estimators/src/PositionFilter.cpp
+++ b/wire_state_estimators/src/PositionFilter.cpp
@@ -37,12 +37,12 @@
 #include "PositionFilter.h"
 #include "KalmanFilter.h"
 
-PositionFilter::PositionFilter() : t_last_update_(0), t_last_propagation_(0), kalman_filter_(0),
-    fixed_pdf_(0), max_acceleration_(0), fixed_pdf_cov_(0), kalman_timeout_(0) {
+PositionFilter::PositionFilter() : t_last_update_(0), t_last_propagation_(0), kalman_filter_(nullptr),
+    fixed_pdf_(nullptr), max_acceleration_(0), fixed_pdf_cov_(0), kalman_timeout_(0) {
 }
 
 PositionFilter::PositionFilter(const PositionFilter& orig) : mhf::IStateEstimator(orig), t_last_update_(orig.t_last_update_),
-    t_last_propagation_(orig.t_last_propagation_), kalman_filter_(0), fixed_pdf_(0), max_acceleration_(orig.max_acceleration_),
+    t_last_propagation_(orig.t_last_propagation_), kalman_filter_(nullptr), fixed_pdf_(nullptr), max_acceleration_(orig.max_acceleration_),
     fixed_pdf_cov_(orig.fixed_pdf_cov_ ), kalman_timeout_(orig.kalman_timeout_) {
 
     if (orig.fixed_pdf_) {
@@ -55,8 +55,10 @@ PositionFilter::PositionFilter(const PositionFilter& orig) : mhf::IStateEstimato
 }
 
 PositionFilter::~PositionFilter() {
-    delete kalman_filter_;
-    delete fixed_pdf_;
+    if (kalman_filter_)
+        delete kalman_filter_;
+    if (fixed_pdf_)
+        delete fixed_pdf_;
 }
 
 PositionFilter* PositionFilter::clone() const {
@@ -88,6 +90,7 @@ void PositionFilter::propagate(const mhf::Time& time) {
             fixed_pdf_->setMean(kalman_filter_->getGaussian().getMean());
         }
 
+        // kalman_filter_ has been checked before, so it isn't a nullptr
         delete kalman_filter_;
         kalman_filter_ = 0;
         return;
@@ -127,11 +130,15 @@ void PositionFilter::update(const pbl::PDF& z, const mhf::Time& time) {
 }
 
 void PositionFilter::reset() {
-    delete kalman_filter_;
-    kalman_filter_ = 0;
+    if (kalman_filter_) {
+        delete kalman_filter_;
+        kalman_filter_ = nullptr;
+    }
 
-    delete fixed_pdf_;
-    fixed_pdf_ = 0;
+    if (fixed_pdf_) {
+        delete fixed_pdf_;
+        fixed_pdf_ = nullptr;
+    }
 }
 
 const pbl::PDF& PositionFilter::getValue() const {
@@ -141,7 +148,7 @@ const pbl::PDF& PositionFilter::getValue() const {
         return *fixed_pdf_;
     }
 
-    std::cout << "SOMETHINGS WRONG" << std::endl;
+    throw std::runtime_error("Couldn't get a value");
 }
 
 bool PositionFilter::setParameter(const std::string& /*param*/, bool /*b*/) {

--- a/wire_state_estimators/src/PositionFilter.cpp
+++ b/wire_state_estimators/src/PositionFilter.cpp
@@ -122,7 +122,7 @@ void PositionFilter::update(const pbl::PDF& z, const mhf::Time& time) {
             kalman_filter_->update(*G);
         }
     } else {
-        printf("PositionFilter can only be updated with Gaussians.\n");
+        std::cout << "PositionFilter can only be updated with Gaussians, not with " << z.type() << std::endl;
     }
 }
 

--- a/wire_state_estimators/src/PositionFilter.cpp
+++ b/wire_state_estimators/src/PositionFilter.cpp
@@ -144,7 +144,7 @@ const pbl::PDF& PositionFilter::getValue() const {
     std::cout << "SOMETHINGS WRONG" << std::endl;
 }
 
-bool PositionFilter::setParameter(const std::string& param, bool b) {
+bool PositionFilter::setParameter(const std::string& /*param*/, bool /*b*/) {
     return false;
 }
 

